### ISSUE TITLE
Cannot reproduce table 2 of the paper with this code

### ISSUE THREAD

**Issue by guglielmocamporese** [BraveGump/CGDL-for-Open-Set-Recognition#10]
I runned the code of this repo on other datasets and this what I obtained. And with this code I'm not able to reproduced the results of Table 2 of the paper:

| Model | MNIST | SVHN | CIFAR10 | CIFAR+10 | CIFAR+50 | TinyImageNet |
| - | - | - | - | - | - | - |
| CGDL from the original paper [here](https://arxiv.org/pdf/2003.08823.pdf) | 0.994 | 0.935 | 0.903 | 0.959 | 0.950 | 0.762 | 
| CGDL from the released code | 0.989 | 0.845 | 0.671 | 0.724 | 0.713 | 0.651 |

**Comment by BraveGump** [comment]
> I runned the code of this repo on other datasets and this what I obtained. And with this code I'm not able to reproduced the results of Table 2 of the paper:
> 
> Model	MNIST	SVHN	CIFAR10	CIFAR+10	CIFAR+50	TinyImageNet
> CGDL from the original paper [here](https://arxiv.org/pdf/2003.08823.pdf)	0.994	0.935	0.903	0.959	0.950	0.762
> CGDL from the released code	0.989	0.845	0.671	0.724	0.713	0.651
> Difference	0.005 (0.5%)	0.090 (9%)	0.232 (23.2%)	0.235 (23.5%)	0.237 (23.7% )	0.111 (11.1%)

May I know how you calculate the AUROC?  For my table, I vary the Gaussian threshold from 0 to 1 with a fixed step and use rectangles to approximate the area to get AUROC, and assume when threshold=0, [TPR, FPR]=[0,0], while when threshold =1, [TPR, FPR]=[1,1],  which is apparent in Mathematics. This way is really different from most previous works (openmax, crosr,c2ae...) as they should just use the AUROC function in the Sklearn package to calculate AUROC.  In their methods, they redistribute probability scores and directly calculate out unknown probability thus they can use the AUROC function in the Sklearn package. However, in our method, we use Gaussian to predict unknowns, thus the AUROC function in the Sklearn package is not applicable in our case. Honestly, this may be not a fair comparison. In my recent work, for this setup, I just evaluate the performance by F1-scores instead of AUROC. 
Hope the above information could help you and so sorry that we did not mention this information in our paper.

**Comment by guglielmocamporese** [comment]
Hi BraveGump,

here my answer:
> May I know how you calculate the AUROC?

I'm using the implementation of AUROC from the [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html) package, it's stable and correct.


> This way is really different from most previous works (openmax, crosr,c2ae...) as they should just use the AUROC function in the Sklearn package to calculate AUROC

If you think your metric is different from AUROC, why you reported as AUROC in the paper? Could you provide your implementation of the AUROC?

> However, in our method, we use Gaussian to predict unknowns, thus the AUROC function in the Sklearn package is not applicable in our case. Honestly, this may be not a fair comparison.

Why it's not applicable in the case of gaussians? For the AUROC you need to output a single scalar as the final prediction. In the gaussian case is the same: you can consider the pdf, the cdf or the kl-divergence.
If you think this may not be a fair comparison, why you did and reported this comparison in the paper?

Best,

Guglielmo

**Comment by BraveGump** [comment]
Thank you for reproduce and pointing out our carelessness in the AUROC experiment.

>why you did and reported this comparison in the paper?

As we did not find released codes of those previous works, when preparing this paper, the method to calculate AUROC is not determined and we did not use the function in sklearn to calculate AUROC. 

>I'm using the implementation of AUROC from the sklearn package, it's stable and correct.

>For the AUROC you need to output a single scalar as the final prediction. In the gaussian case is the same: you can consider the pdf, the cdf or the kl-divergence.

Your method to calculate AUROC is more reasonable. In our recent work, we will report this and follow this to calculate AUROC.

Thank you again!

**Comment by guglielmocamporese** [comment]
Hi BraveGump,

> As we did not find released codes of those previous works, when preparing this paper, the method to calculate AUROC is not determined and we did not use the function in sklearn to calculate AUROC.

I think that the point is not in the implementation you used (custom or from sklearn), the point is on the fairness of the experiment and protocol: is the AUROC  really the function that you implemented in this paper? If yes the table in your paper is fair, otherwise, it is not, you measured something else, sorry.

> Your method to calculate AUROC is more reasonable. In our recent work, we will report this and follow this to calculate AUROC.

It is not _more_ reasonable, it is the only correct way for computing the AUROC (you can implement by your own for sure, but it has to output the same thing). The problem is that other researchers are working on this, and you are saying that you have reported an unfair, and strong result on your paper (from our experiments with your code, your results changed a LOT, ~23% in one case).

Best, 

Guglielmo
